desc/efit.py

```diff
@@ -0,0 +1,374 @@
+from matplotlib import pyplot as plt
+from scipy.interpolate import CubicSpline
+
+from desc.basis import DoubleFourierSeries
+from desc.equilibrium import Equilibrium
+from desc.geometry import FourierRZToroidalSurface
+from desc.grid import LinearGrid
+from desc.profiles import PowerSeriesProfile, SplineProfile
+from desc.transform import Transform
+
+
+class EFITIO:
+    """Performs input from EFIT files to DESC Equilibrium."""
+
+    @classmethod
+    def load(  # noqa: C901 - fxn too complex
+        cls,
+        g,
+        M,
+        N=0,
+        L=None,
+        bdry_dist=1,
+    ):
+        """Create a DESC equilibrium object from an EFIT geqdsk file.
+
+        The equilibrium will not be solved, this just creates an equilibirum
+        with the correct boundary and profiles.
+
+        Parameters
+        ----------
+        g : path-like or dict
+            Path to geqdsk file, or dict of already read parameters.
+        M : int
+            Poloidal resolution of the equilibrium. Used for fitting the boundary.
+        N : int, optional
+            Toroidal resolution of the equilibrium.
+        L : int, optional
+            Radial resolution of the equilibrium. Used for fitting profiles.
+            If None, spline profiles will be used. Otherwise they are fit
+            to even power series in rho.
+        bdry_dist : float
+            Deviation from separatrix. If bdry_dist == 1, then the boundary
+            is the actual separatrix read from the geqdsk file.
+            bdry_dist = 0 corresponds to the magnetic axis.
+            Typical values should be 0.95 < bdry_dist < 1.0
+
+        Returns
+        -------
+        eq : Equilibrium
+            DESC equilibrium with boundary shape, pressure, iota, and
+            total toroidal flux from geqdsk file.
+        """
+        if not isinstance(g, dict):
+            with open(g) as f:
+                lines = f.readlines()
+            g = {}
+
+            has_spaces = all(
+                [elem == 5 for elem in [len(lines[i].split()) for i in range(1, 5)]]
+            )
+            if has_spaces:
+
+                def splitline(s):
+                    return s.split()
+
+                def splitarr(l):
+                    return " ".join(l).split()
+
+            else:
+
+                def splitline(s):
+                    ss = s.split("\n")[0]
+                    idx = np.arange(0, len(ss) + 1, 16)
+                    return [ss[i:j] for i, j in zip(idx[:-1], idx[1:])]
+
+                def splitarr(l):
+                    return splitline("".join([line.split("\n")[0] for line in l]))
+
+            line0 = lines[0].split()
+
+            # points in the horizontal direction (width) and vertical direction (height)
+            g["nw"] = int("".join([c for c in line0[-2] if c.isdigit()]))
+            g["nh"] = int("".join([c for c in line0[-1] if c.isdigit()]))
+            if len(line0) == 8:
+                g["efit"] = line0[4]
```

Comment:
what is this key? wouuld be good to say what this is (is it the EFIT version?)

Reply:
It's the efit file information. I have added a comment. It doesn't affect the calculation in any way.